src/Init/Tactics.lean

```diff
@@ -941,6 +1008,61 @@ If there are several with the same priority, it is uses the "most recent one". E
 ```
 -/
 syntax (name := simp) "simp" (Tactic.simpPre <|> Tactic.simpPost)? (ppSpace prio)? : attr
+
+
+/-- The possible `norm_cast` kinds: `elim`, `move`, or `squash`. -/
+syntax normCastLabel := &"elim" <|> &"move" <|> &"squash"
+
+/--
+The `norm_cast` attribute should be given to lemmas that describe the
+behaviour of a coercion in regard to an operator, a relation, or a particular
+function.
+
+It only concerns equality or iff lemmas involving `↑`, `⇑` and `↥`, describing the behavior of
+the coercion functions.
+It does not apply to the explicit functions that define the coercions.
+
+Examples:
+```lean
+@[norm_cast] theorem coe_nat_inj' {m n : ℕ} : (↑m : ℤ) = ↑n ↔ m = n
+
+@[norm_cast] theorem coe_int_denom (n : ℤ) : (n : ℚ).denom = 1
+
+@[norm_cast] theorem cast_id : ∀ n : ℚ, ↑n = n
+
+@[norm_cast] theorem coe_nat_add (m n : ℕ) : (↑(m + n) : ℤ) = ↑m + ↑n
+
+@[norm_cast] theorem cast_coe_nat (n : ℕ) : ((n : ℤ) : α) = n
+
+@[norm_cast] theorem cast_one : ((1 : ℚ) : α) = 1
+```
+
+Lemmas tagged with `@[norm_cast]` are classified into three categories: `move`, `elim`, and
+`squash`. They are classified roughly as follows:
+
+* elim lemma:   LHS has 0 head coes and ≥ 1 internal coe
+* move lemma:   LHS has 1 head coe and 0 internal coes,    RHS has 0 head coes and ≥ 1 internal coes
+* squash lemma: LHS has ≥ 1 head coes and 0 internal coes, RHS has fewer head coes
```

Comment:
```suggestion
* `elim` lemma:   the LHS does not have `coe` at its head, and contains ≥ 1 non-head `coe`
* `move` lemma:  the LHS has `coe` at its head, and no internal occurrences, while the RHS is not headed by `coe` and contains ≥ 1 internal `coe`s
* `squash` lemma: LHS has ≥ 1 `coe`s at its head and 0 internal `coe`s, RHS has fewer head `coe`s

In other words, `elim` lemmas are used to eliminate occurrences of `coe`, `move` lemmas are used to push `coe`s inwards, and `squash` lemmas are used to simplify stacked coercions.
```
If this is inaccurate (e.g. "coe" in the original may have referred to all three types of coercions mentioned earlier), let me know and I'll suggest alternate phrasing.